migrations/capcons/capabilitymigration.go

```diff
@@ -0,0 +1,116 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package capcons
+
+import (
+	"github.com/onflow/cadence/migrations"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/errors"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/sema"
+)
+
+type CapabilityMigrationReporter interface {
+	MigratedPathCapability(
+		accountAddress common.Address,
+		addressPath interpreter.AddressPath,
+		borrowType *interpreter.ReferenceStaticType,
+	)
+	MissingCapabilityID(
+		accountAddress common.Address,
+		addressPath interpreter.AddressPath,
+	)
+}
+
+// CapabilityMigration migrates all path capabilities to ID capabilities,
+// using the path to ID capability controller mapping generated by LinkMigration.
+type CapabilityMigration struct {
+	CapabilityIDs map[interpreter.AddressPath]interpreter.UInt64Value
+	Reporter      CapabilityMigrationReporter
+}
+
+var _ migrations.Migration = &CapabilityMigration{}
+
+func (*CapabilityMigration) Name() string {
+	return "CapabilityMigration"
+}
+
+var fullyEntitledAccountReferenceStaticType = interpreter.ConvertSemaReferenceTypeToStaticReferenceType(
+	nil,
+	sema.FullyEntitledAccountReferenceType,
+)
+
+// Migrate migrates a path capability to an ID capability in the given value.
+// If a value is returned, the value must be updated with the replacement in the parent.
+// If nil is returned, the value was not updated and no operation has to be performed.
+func (m *CapabilityMigration) Migrate(
+	addressPath interpreter.AddressPath,
+	value interpreter.Value,
+	_ *interpreter.Interpreter,
+) interpreter.Value {
+	reporter := m.Reporter
+
+	switch value := value.(type) {
+	case *interpreter.PathCapabilityValue: //nolint:staticcheck
+
+		// Migrate the path capability to an ID capability
+
+		oldCapability := value
+
+		capabilityAddressPath := oldCapability.AddressPath()
+		capabilityID, ok := m.CapabilityIDs[capabilityAddressPath]
+		if !ok {
+			if reporter != nil {
+				reporter.MissingCapabilityID(
+					addressPath.Address,
+					capabilityAddressPath,
+				)
+			}
+			break
+		}
+
+		newBorrowType, ok := oldCapability.BorrowType.(*interpreter.ReferenceStaticType)
+		if !ok {
+			panic(errors.NewUnreachableError())
+		}
+
+		// Convert the old AuthAccount type to the new fully-entitled Account type
+		if newBorrowType.ReferencedType == interpreter.PrimitiveStaticTypeAuthAccount { //nolint:staticcheck
+			newBorrowType = fullyEntitledAccountReferenceStaticType
+		}
```

Comment:
True, in the full migration pipeline this should not be encountered, but it helps with keeping the migration and its test independent

Reply:
Makes sense. Just to flag: here it might not catch if the Account type is nested inside, for e.g: optional auth account i.e: `AuthAccount?` (not sure if that's possible even)